### PR TITLE
change json schema transform from `replace` to `return`

### PR DESCRIPTION
**Pull request by OsmondJiang**
avoid replacing existing model, which is not thread safe.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4767)

### CODE REVIEW

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 1]
```diff
@@ -63,70 +63,91 @@ public JsonSchemaTransformer(JsonSchema schema)
...
+                JsonUtility.SetSourceInfo(transformedScalar, JsonUtility.GetSourceInfo(node));
+                return (default, transformedScalar);
             });
         }
 
-        private void Traverse(JsonSchema schema, JToken token, Func<JsonSchema, JToken, string[]> transform)
+        private JToken Traverse(JsonSchema schema, JToken token, Func<JsonSchema, JToken, (string[], JToken)> transform)
         {
             schema = _definitions.GetDefinition(schema);
+            var sourceInfo = JsonUtility.GetSourceInfo(token);
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: To make things simpler, the transformed token does not need to preserve source info, there is no downstream consumers that needs source info afterwards. Same thing can be removed in  line 183.

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 2]
```diff
@@ -63,70 +63,91 @@ public JsonSchemaTransformer(JsonSchema schema)
...
             if (schema == null)
             {
-                return;
+                return token;
             }
 
-            var transformedKeys = transform(schema, token) ?? Array.Empty<string>();
+            var (transformedKeys, transformedToken) = transform(schema, token);
+            transformedKeys = transformedKeys ?? Array.Empty<string>();
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: can compare `token` and `transformedToken` for reference equality to detect if the token has been processed. If the token is processed, skip the entire switch block in line 107.
With this, the callback does not need to return `transformedKeys`

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 3]
```diff
@@ -63,70 +63,91 @@ public JsonSchemaTransformer(JsonSchema schema)
...
 
-            switch (token)
+            switch (transformedToken)
             {
+                case JValue scalar:
+
+                    return scalar;
+
                 case JArray array:
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: There is no need to copy the array if `schema.Items == null`:
```suggestion
                case JArray array when schema.Items != null:
```

With this, can replace line 109 ~ line 112 and line 150 with a default block that return `token`

File: src/docfx/lib/schema/JsonSchemaTransformer.cs (hunk 1 above)
Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: I am not sure about that, will xref need that?

File: src/docfx/lib/schema/JsonSchemaTransformer.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: good point!

File: src/docfx/lib/schema/JsonSchemaTransformer.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: when the token is process, if the token is JObject, maybe only a few properties are process, so reference equality is not working here

File: src/docfx/lib/schema/JsonSchemaTransformer.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: nop, after schema transform, it is turned into a json object and pass to template

Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: @live1206  confirmed it's not used in xref, so delete it

**Pull request merged**